Added discovery design doc (#189)
redhat-et/apex: docs/design/design-discovery.md

- The goal is to create direct peering between nodes where possible and bounce connections through a relay node where direct connections are not possible.

```
                      To Internet

                          |
                          |
                          |  /------------  Relayed
                      Y:y | /               Address
                      +--------+
                      |        |
                      |  TURN  |
                      | Server |
                      |        |
                      +--------+
                          |
                          |
                          | /------------  Server
                   X1':x1'|/               Reflexive
                    +------------+         Address
                    |    NAT     |
                    +------------+
                          |
                          | /------------  Local
                      X:x |/               Address
                      +--------+
                      |        |
                      | Agent  |
                      |        |
                      +--------+

```

> Figure 1. *Candidate Types*

Reviewer: ```suggestion
*Figure 1. Candidate Types*
```

> ### Current and Future Discovery Plans

Reviewer: ```suggestion
## Current and Future Discovery Plans
```

> Disclaimer: current peer discovery is still in an early POC state. 

Reviewer: Github has special rendering support for a `Note` or `Warning` in this format.

**Nodes without a firewall and/or NAT device between them (currently supported):**

> - Apex will peer end nodes directly that share the same STUN IP address. 

Reviewer: actually, you use "reflexive address" in the diagram and the text on the next line. Maybe that's better? You could add a subsection here that explains how the reflexive address is determined. That could be where you discuss STUN.

> - Apex will peer end nodes directly that share the same STUN IP address. 

Author: Added more details on STUN in the next section. Thanks!

> - This means when both nodes share the same public or "reflexive address" (as defined in the above figure) then Apex makes an assumption that other peers are likely to have direct access to one another. Each peer has their stun/reflexive address in the peer listing received from the controller.

Reviewer: ```suggestion
- This means that when both nodes share the same public or "reflexive address" (as defined in the above figure), Apex assumes that other peers are likely to have direct access to one another. Each peer has their stun/reflexive address in the peer listing received from the controller.
```

> - Next the Apex agent will look up the "Local Address" (see Figure 1) of a peer candidate in the peer listing and attempt to probe for connectivity. If this probing succeeds, we consider it a likely candidate match and both peers setup the connection to one another with a /32 host route and wireguard tunnel.

Reviewer: ```suggestion
- Next, the Apex agent will look up the "Local Address" (see Figure 1) of a peer candidate in the peer listing and attempt to probe for connectivity. If this probing succeeds, we consider it a likely candidate match, and both peers set up the connection to one another with a /32 host route and wireguard tunnel.
```

**Nodes with a firewall and/or NAT device between them (currently in progress):**

```
                               +---------+
             +--------+        |Signaling|         +--------+
             | STUN   |        |Server   |         | STUN   |
             | Server |        +---------+         | Server |
             +--------+       /           \        +--------+
                             /             \
                            /               \
                           / <- Signaling -> \
                          /                   \
                   +--------+               +--------+
                   |  NAT   |               |  NAT   |
                   +--------+               +--------+
                      /                             \
                     /                               \
                 +-------+                       +-------+
                 | Agent |                       | Agent |
                 |   L   |                       |   R   |
                 +-------+                       +-------+

> ```

Reviewer: Do you want to show something about how the wireguard traffic flows in this case? I see the `Signaling` label, but it's not clear that the data traffic flows any differently. That seems like an important distinction.

Author: Renamed to relay and changed the ASCII art.

> Figure 2 - NAT traversal (we are refering to the signalling server here as a relay node)

Reviewer: ```suggestion
*Figure 2 - NAT traversal (we are referring to the signaling server here as a relay node)*
```

- Two endpoints that do not match the initial local address peering, will attempt to peer via the STUN method (see the ICE RFC for further STUN details). A STUN request can be used to open a source UDP port on the NAT device front-ending the endpoint.
- That UDP source port will remain open for some unknown period of time (depending on the NAT device).
- Next we attempt to match peer candidates with that IP:Port STUN address for the endpoint.
- All nodes will attach to a relay node running a STUN server and/or terminating wireguard peerings for fallback forwarding in the case a peer candidate match cannot be made. This peering can be reconciled from wireguard state until we no longer terminate peers on the relay node (tbd).
- This peering can handle multiple nodes behind a PAT (port address translation) device since there will always be a unique source address assigned in the connection state.

> **Nodes that still cannot find a match (currently supported):**

Reviewer: ```suggestion
### Nodes that still cannot find a match (currently supported)
```

> - For all nodes that are unable to find a suitable candidate peering, we fallback to forwarding traffic through the relay node that need to be open (UDP 51820) to all endpoints in the zone.

Reviewer: ```suggestion
- For all nodes that are unable to find a suitable candidate peering, we fall back to forwarding traffic through the relay node that needs to be open (UDP 51820) to all endpoints in the zone.
```

> - Currently all nodes that peer into a zone, will peer to the relay node to ensure connectivty in all scenarios. Since this is a hairpin, it is the least ideal pairing but can be scaled to fit the small amount of peers that cannot establish a direct peering either through the "Local Address" or STUN discovered "Reflexive Address".

Reviewer: ```suggestion
- Currently, all nodes that peer into a zone will peer to the relay node to ensure connectivity in all scenarios. Since this is a hairpin, it is the least ideal pairing but can be scaled to fit the small number of peers that cannot establish a direct peering either through the "Local Address" or STUN discovered "Reflexive Address".
```

- (Future work) The relay node will likely be the point at which source ports are opened and reconciled against in peer listings.

> - (Future work) Eliminating the need to terminate wireguard sessions to the relay node would reduce the attack surface and better adhere to zero trust networking principles.

Reviewer: ```suggestion
- (Future work) Eliminating the need to terminate wireguard sessions to the relay node would reduce the attack surface and better adhere to zero trust networking principles. [Issue 169](https://github.com/redhat-et/apex/issues/169)
```

**Health checking and proper probing (future work)**

- Current probing is remedial at best. It needs to probe for a likely wireguard pairing and not just general connectivity even though we know the peer is running the agent and presume there is a wireguard port listening.
- Ideally we probe for a crypto pairing capability.  

> - Constant Health checking will be required to enable for nodes to flip to different peerings as underlying topological changes occur. For example, if the STUN socket between two peers is no longer available, the two peer nodes should fallback to the relay node to enable continuity in the connection. Vice, versa, if a better peering candidacy returns, the nodes should flip back to the direct peering either through the local address or reflexive address (Figure 1).

Reviewer: ```suggestion
- Constant Health checking will be required to enable nodes to flip to different peerings as underlying topological changes occur. For example, if the STUN socket between two peers is no longer available, the two peer nodes should fall back to the relay node to enable continuity in the connection. Vice, versa, if a better peering candidacy returns, the nodes should flip back to direct peering either through the local address or reflexive address (Figure 1).
```

**Who determines peer candidates?**

> - Whether the controller determines peering matches or provides enough information about peers to the agent to determine proper candidacy is still tbd. Ideally the controller is not in the process outside of peer listing updates allowing the agents to negotiate peering on the fly.

Reviewer: ```suggestion
- Whether the controller determines peering matches or provides enough information about peers to the agent to determine proper candidacy is still tbd. Ideally, the controller is not in the process outside of peer listing updates allowing the agents to negotiate peering on the fly.
```
